### DISCUSSION TITLE
Problem with monai.data.utils.iter_patch

### DISCUSSION THREAD

**Discussion by Mo-Junyang** [Project-MONAI/MONAI#6606] [Q&A] [answered]
Firstly, I want to crop some overlap 3D patches with sliding window from a medical Image. Then I found the function iter_patch and I checked if it work in a simple way.  
```python
    a = np.array([[1,2,3,4],[5,6,7,8],[8,7,6,5],[4,3,2,1]])
    print(a)
    c= iter_patch(a,patch_size=[2,2],start_pos=0,overlap=0.5,copy_back=False)
    for i,b in c:
        print(i,"\n")
``` 

but get a wired output like this

    [[1 2 3 4]
     [5 6 7 8]
     [8 7 6 5]
     [4 3 2 1]]
    [[8 7]
     [4 3]] 
 
    [[7 6]
     [3 2]] 

    [[6 5]
     [2 1]] 

    [[5 8]
     [1 4]] 

    [[4 3]] 

    [[3 2]] 

    [[2 1]] 

    [[1 4]] 

    [] 

    [] 

    [] 

    [] 

    [] 

    [] 

    [] 

    [] 

but when I change the Input into string, it seems to work well
```python
    a = np.array([["a", "a", "b", "b"], ["a", "a", "b", "b"], ["c", "c", "d", "d"], ["c", "c", "d", "d"]])
    print(a)
    c= iter_patch(a,patch_size=[2,2],start_pos=0,overlap=0.5,copy_back=False)
    for i,b in c:
        print(i,"\n")
``` 

```
[['a' 'a' 'b' 'b']
 ['a' 'a' 'b' 'b']
 ['c' 'c' 'd' 'd']
 ['c' 'c' 'd' 'd']]
[['a' 'a']
 ['a' 'a']] 

[['a' 'b']
 ['a' 'b']] 

[['b' 'b']
 ['b' 'b']] 

[['b' 'a']
 ['b' 'a']] 

[['a' 'a']
 ['c' 'c']] 

[['a' 'b']
 ['c' 'd']] 

[['b' 'b']
 ['d' 'd']] 

[['b' 'a']
 ['d' 'c']] 

[['c' 'c']
 ['c' 'c']] 

[['c' 'd']
 ['c' 'd']] 

[['d' 'd']
 ['d' 'd']] 

[['d' 'c']
 ['d' 'c']] 

[['c' 'c']
 ['a' 'a']] 

[['c' 'd']
 ['a' 'b']] 

[['d' 'd']
 ['b' 'b']] 

[['d' 'c']
 ['b' 'a']] 

``` 
So, I want to know what happened, and how can I create some overlap patches and save it in files with monai? Thanks a lot

**Comment by KumoLiu** [comment, marked as answer]
Hi @Mo-Junyang, you need to ensure the data is channel-first, try something like this:
```
a = np.array([[[1,2,3,4],[5,6,7,8],[8,7,6,5],[4,3,2,1]]])
print(a)
c= iter_patch(a,patch_size=[None,2,2],start_pos=(0,0),overlap=0,copy_back=False)
for i,b in c:
    print(i,"\n")
```
You can also try to use `PatchDataset` and `GridPatchDataset`

Hope it can help you, thanks!

**Reply by Mo-Junyang** [reply]
Thank you! That helps a lot! BTW, do you know why it didn't go wrong when the input was str type array, even though I didn't ensure the "channel-first"

**Reply by KumoLiu** [reply]
Sure, because of the different dtype, a different backend is used.
https://github.com/Project-MONAI/MONAI/blob/a1c437111a511d17dcd0e04d491747f6677f4def/monai/transforms/croppad/functional.py#L99
Thanks!

**Reply by Mo-Junyang** [reply]
Thanks a lot, it makes me clearer！
